src/exporter/DaemonMetricCollector.cc

```diff
@@ -0,0 +1,163 @@
+  }
+  else if (value.is_double()) {
+    double v = value.as_double();
+    add_metric(ss, v, name, description, mtype, labels);
+  }
+}
+
+std::string boost_string_to_std(boost::json::string js) {
+  std::string res(js.data());
+  return res;
+}
+
+std::string quote(std::string value) {
+	return "\"" + value + "\"";
+}
+
+void DaemonMetricCollector::send_requests() {
+  std::string result;
+  for(auto client : clients) {
+    AdminSocketClient &sock_client = client.second;
+    std::string daemon_name = client.first;
+    std::string request("{\"prefix\":\"perf dump\"}");
+    std::string response;
+    sock_client.do_request(request, &response);
+    if (response.size() > 0) {
+      json_object dump = boost::json::parse(response).as_object();
+      request = "{\"prefix\":\"perf schema\"}";
+      response = "";
+      sock_client.do_request(request, &response);
+      json_object schema = boost::json::parse(response).as_object();  
+      for (auto perf : schema) {
+        std::string perf_group = perf.key().to_string();
+        json_object perf_group_object = perf.value().as_object();
+        for (auto perf_counter : perf_group_object) {
+          std::string perf_name = perf_counter.key().to_string();
+          json_object perf_info = perf_counter.value().as_object();
+          if (perf_info["priority"].as_int64() < PerfCountersBuilder::PRIO_USEFUL) {
+            continue;
+          }
+          int64_t type = perf_info["type"].as_int64();
+          std::string mtype = boost_string_to_std(perf_info["metric_type"].as_string());
+          std::string description = boost_string_to_std(perf_info["description"].as_string());
+          std::stringstream ss;
+          std::string name = "ceph_" + perf_group + "_" + perf_name;
+
+          std::string labels;
+          // Labels
+          // FIXME: test this, based on mgr_module perfpath_to_path_labels
+          if (daemon_name.find("rgw") != std::string::npos) {
+            labels = std::string("instance_id=") + quote(daemon_name.substr(4, std::string::npos));
+          } else {
+            labels = "ceph_daemon=" + quote(daemon_name);
+            if (daemon_name.find("rbd-mirror") != std::string::npos) {
+              std::regex re("^rbd_mirror_image_([^/]+)/(?:(?:([^/]+)/)?)(.*)\\.(replay(?:_bytes|_latency)?)$");
+              std::smatch match;
+              if (std::regex_search(daemon_name, match, re) == true) {
+                name = "ceph_rbd_mirror_image_" + match.str(4);
+                labels += ",pool=" + quote(match.str(1));
+                labels += ",namespace=" + quote(match.str(2));
+								labels += ",image=" + quote(match.str(3));
+              }
+            }
+          }
+
+          // value and add_metric
+          json_value perf_values = dump[perf_group].as_object()[perf_name];
+          if (type & PERFCOUNTER_LONGRUNAVG) {
+            int64_t count = perf_values.as_object()["avgcount"].as_int64();
```

Comment:
@epuertat I did some refactoring on that. Since this function is more complex I'm not sure I like extracting to a lot of functions because they aren't being used more than one time. Anyways, I did it.

Reply:
Reuse is only one of the reasons to extract code to an external method: improving code quality (readability, cyclomatic complexity) or testability are good reasons too.

That code comprises a triple-nested for loop, so breaking that down to separate methods would allow you to improve unit test isolation (testing the inner methods without the overhead of outer iterations; testing the outer methods mocking the inner ones).

Reply:
Yeah I understand that. I'll add a new task to refactor this on the exporter doc so that we keep track of it. But for now, let's try to push this PR onto the feature branch.